cmd/server-main.go

```diff
@@ -306,42 +297,73 @@ func serverMain(ctx *cli.Context) {
 		logger.FatalIf(err, "Unable to initialize backend")
 	}
 
-	globalObjLayerMutex.Lock()
-	globalObjectAPI = newObject
-	globalObjLayerMutex.Unlock()
-
 	// Populate existing buckets to the etcd backend
 	if globalDNSConfig != nil {
-		initFederatorBackend(newObject)
+		initFederatorBackend(newObjectLayerFn())
 	}
 
+	logger.FatalIf(initPolicySys(), "Unable to initialize policy system")
+
+	// Re-enable logging
+	logger.Disable = false
+
+	logger.LogIf(context.Background(), initNotificationSys())
+
 	// Prints the formatted startup message once object layer is initialized.
 	apiEndpoints := getAPIEndpoints(globalMinioAddr)
 	printStartupMessage(apiEndpoints)
 
 	// Set uptime time after object layer has initialized.
 	globalBootTime = UTCNow()
 
-	// Re-enable logging
-	logger.Disable = false
-
 	handleSignals()
 }
 
+func initPolicySys() error {
+	// Create new policy system.
+	globalPolicySys = NewPolicySys()
+
+	// Initialize policy system.
+	if err := globalPolicySys.Init(newObjectLayerFn()); err != nil {
+		return uiErrUnableToReadFromBackend(err).Msg("Unable to initialize policy system")
+	}
+
+	return nil
+}
+
+func initNotificationSys() error {
+	// Create new notification system.
+	globalNotificationSys = NewNotificationSys(globalServerConfig, globalEndpoints)
+	// Initialize notification system.
+	return globalNotificationSys.Init(newObjectLayerFn())
+}
+
 // Initialize object layer with the supplied disks, objectLayer is nil upon any error.
-func newObjectLayer(endpoints EndpointList) (newObject ObjectLayer, err error) {
-	// For FS only, directly use the disk.
+func initObjectLayer(endpoints EndpointList) error {
 
+	var newObject ObjectLayer
+	var err error
+
+	// For FS only, directly use the disk.
 	isFS := len(endpoints) == 1
 	if isFS {
 		// Initialize new FS object layer.
-		return NewFSObjectLayer(endpoints[0].Path)
+		newObject, err = NewFSObjectLayer(endpoints[0].Path)
+	} else {
+		format, formatErr := waitForFormatXL(context.Background(), endpoints[0].IsLocal, endpoints, globalXLSetCount, globalXLSetDriveCount)
+		if formatErr != nil {
+			return formatErr
+		}
+		newObject, err = newXLSets(endpoints, format, len(format.XL.Sets), len(format.XL.Sets[0]))
 	}
 
-	format, err := waitForFormatXL(context.Background(), endpoints[0].IsLocal, endpoints, globalXLSetCount, globalXLSetDriveCount)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	return newXLSets(endpoints, format, len(format.XL.Sets), len(format.XL.Sets[0]))
+	globalObjLayerMutex.Lock()
+	globalObjectAPI = newObject
```

Comment:
Why are you making all the unnecessary changes?  avoid making global API edits, just keep them as they are in server-main.go.

Reply:
> Why are you making all the unnecessary changes? avoid making global API edits, just keep them as they are in server-main.go.

I just wanted to make serverMain() function little shorter, it is already long

Reply:
My reasoning is that let's avoid code relocation in this simple change @vadmeste unless you cannot avoid it. I clearly see this functionality to be independent of such changes.

Reply:
> My reasoning is that let's avoid code relocation in this simple change @vadmeste unless you cannot avoid it. I clearly see this functionality to be independent of such changes.

Right, I'll revert it